numba/tests/test_function.py

```diff
@@ -0,0 +1,575 @@
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    f = cfunc(sig)(func)
+    f.pyfunc = func
+    return f
+
+
+def njit_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    f = njit(func)
+    f.pyfunc = func
+    return f
+
+
+def njit2_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    f = njit(sig)(func)
+    f.pyfunc = func
+    return f
+
+
+def ctypes_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
+    cfunc = cfunc_func(func, sig)
+    addr = cfunc._wrapper_address
+    if sig == int64(int64):
+        f = ctypes.CFUNCTYPE(ctypes.c_int64)(addr)
+        f.pyfunc = func
+        return f
+    raise NotImplementedError(
+        f'ctypes decorator for {func} with signature {sig}')
+
+
+class WAP(types.WrapperAddressProtocol):
+
+    def __init__(self, func, sig):
+        self.pyfunc = func
+        self.cfunc = cfunc_func(func, sig)
+        self.sig = sig
+
+    def __wrapper_address__(self, sig):
+        assert self.sig == sig, (self.sig, sig)
+        return self.cfunc._wrapper_address
+
+    def signature(self):
+        return self.sig
+
+    def __call__(self, *args, **kwargs):
+        return self.pyfunc(*args, **kwargs)
+
+
+def wap_func(func, sig=int64(int64)):
+    return WAP(func, sig)
+
+
+all_func_kinds = [pure_func, cfunc_func, njit_func,
+                  njit2_func, ctypes_func, wap_func]
+
+
+class TestFuncionType(TestCase):
+
+    def _test_issue_3405_original(self):
+
+        @njit
+        def a():
+            return 2
+
+        @njit
+        def b():
+            return 3
+
+        def g(arg):
+            if arg:
+                f = a
+            else:
+                f = b
+            out = f()
+            return out
+
+        self.assertEqual(g(True), 2)
+        self.assertEqual(g(False), 3)
+
+        self.assertEqual(njit(g)(True), g(True))
+        self.assertEqual(njit(g)(False), g(False))
+
+    def test_issue_3405_using_cfunc(self):
+
+        @cfunc('int64()')
+        def a():
+            return 2
+
+        @cfunc('int64()')
+        def b():
+            return 3
+
+        def g(arg):
+            if arg:
+                f = a
+            else:
+                f = b
+            out = f()
+            return out
+
+        self.assertEqual(njit(g)(True), 2)
+        self.assertEqual(njit(g)(False), 3)
+
+    def test_pr4967_example(self):
+
+        @cfunc('int64(int64)')
+        def a(i):
+            return i + 1
+
+        @cfunc('int64(int64)')
+        def b(i):
+            return i + 2
+
+        @njit
+        def foo(f, g):
+            i = f(2)
+            seq = (f, g)
+            for fun in seq:
+                i += fun(i)
+            return i
+
+        a_ = a._pyfunc
+        b_ = b._pyfunc
+        self.assertEqual(foo(a, b),
+                         a_(2) + a_(a_(2)) + b_(a_(2) + a_(a_(2))))
```

Comment:
I think it would be good to move these to a separate `TestMiscIssues` test class to differentiate them from tests defined by design. I would imagine there may be more user issues in the future, quite a few features have a "user issues" test class like this already.

Reply:
I agree. Fixed.